### PR TITLE
Interfaces

### LINKED ISSUE #405: Can not visist when use default interface option
I use a win10 pc, when I  used the default option

> miniserve

it shows
> Serving path \\?\D:\Tool\bin at http://[::]:8080, http://127.0.0.1:8080

my inner ip is 192.168.3.103 when I visit the web it can not show anything. When I use the set the interface 

> miniserve -i 192.168.3.103

it works.it seems the [::] can not work in my windows

### LINKED ISSUE #422: [feature suggestion] show the IP address from all interfaces
miniserve could print the IP address from all interfaces it's listening on:

*http-server*
```
lvmc@Luizs-MacBook-Pro ~ % http-server
Starting up http-server, serving ./public
Available on:
  http://127.0.0.1:8081
  http://192.168.0.101:8081
  http://192.168.15.7:8081
Hit CTRL-C to stop the server
```

*miniserve*
```
lvmc@Luizs-MacBook-Pro Desktop % miniserve -q -v .
miniserve v0.10.3
00:32:55 [INFO] Starting 16 workers
00:32:55 [INFO] Starting "actix-web-service-[::]:8080" service on [::]:8080
00:32:55 [INFO] Starting "actix-web-service-127.0.0.1:8080" service on 127.0.0.1:8080
Serving path /Users/lvmc/Desktop at http://[::]:8080, http://127.0.0.1:8080

Quit by pressing CTRL-C
```

It helps a lot to share the links on local networks.

### LINKED ISSUE #470: [feature request]add 0.0.0.0 to bind to all interfaces
Add support for using 0.0.0.0 to bind to all interfaces.

If my pc have 2 ip address, 192.168.1.111 and 192.168.2.222, i appreciate the program find all the ip address automatically and bind to them when i using the following command.

miniserve -i 0.0.0.0 /tmp/myshare

### PR DESCRIPTION
**Pull request by aliemjay**
## 1. Report if any address fails to bind
`::actix_web::HttpServer::bind()` fails if and only if ALL the requested addresses fail to bind! an the docs require multiple calls for multiple addresses.

## 2. Fix: faulty bahvior when setting `-i 0.0.0.0`
Currrently, `0.0.0.0` address is converted into `127.0.0.1`!

Closes #470

## 3. Fix: On Windows, miniserve binds only to localhost(127.0.0.1) by default
This is in part because of the issue mentioned above, and also because binding to `::` binds only to IPv6 by default on Windows!

Closes #405 
Also reported by @KaneGreen in #470.

## 4. cross-platform default binding addresses
Currently, miniserve by default binds to both `0.0.0.0` and `::` ant the same time. This behaviour has a problem:
> On some platforms (e.g. majority of linux distros and OSX?) ,  the `::` address also accepts ipv4 connections. This makes these two addresses incompatible!
See https://stackoverflow.com/questions/1618240/how-to-support-both-ipv4-and-ipv6-connections

~I currently have no idea how this should be fixed!~

## 5. implement #422
Closes #422

### CODE REVIEW

**Comment by svenstaro** [comment]
Hey, so just to be clear: this is currently wip or shall I have a look?

**Comment by aliemjay** [comment]
This is in progress. I made it a draft.
You can still review the first 3 points. I think they are finished.

File: src/main.rs [hunk 1]
```diff
@@ -378,6 +363,18 @@ async fn run(miniserve_config: MiniserveConfig) -> Result<(), ContextualError> {
         .map_err(|e| ContextualError::IoError("".to_owned(), e))
 }
 
+// Allows us to set low-level socket options:
```
Comment (svenstaro) [inline src/main.rs:366]: Let's use proper docstrings (`///`) for documenting functions.

Comment (aliemjay) [reply src/main.rs:366]: Okay.

What do you think about this overall?
I want to settle this to push #422 fix

Comment (svenstaro) [reply src/main.rs:366]: I really like it. I'm just skeptical about cross-platform support which is very important to me. This needs to work the same on all operating systems (which arguably currently it doesn't really, especially on Windows).

**Comment by aliemjay** [comment]
> I really like it. I'm just skeptical about cross-platform support which is very important to me. This needs to work the same on all operating systems (which arguably currently it doesn't really, especially on Windows).

Well, it should be cross platform: the `IPv6_ONLY` sockopt is a standard and the whole function is adapted from actix-web:
https://github.com/actix/actix-web/blob/f462aaa7b6129b633fa53935d0ee54dbc01b70ea/src/server.rs#L609

**Comment by aliemjay** [comment]
Can we avoid binding to "::" altogether an bind only to "0.0.0.0" by default?
In this case it would be simpler and we should then drop https://github.com/svenstaro/miniserve/pull/500/commits/2a28e6b31e7773debdfd8983e62496cd4af3b9f4 as it would be no longer neccessary.

**Comment by svenstaro** [comment]
I believe on Windows that would result in it not binding to IPv6 at all (but that would have to be tested). Perhaps it might be better to bind to every interface directly individually instead of using `0.0.0.0/::`. The current situation isn't great so I'm happy for alternative implementations.

**Comment by aliemjay** [comment]
Binding to every interface individually is the worst option here:
* First, there is no portable cross-platform way to enumerate all local interfaces at all.
* Second, it has a different behaviour to binding to `0.0.0.0/::` - changing the network configuration, for example, or adding a new interface may render the server inaccessible because the bindings do not update accordingly. 

This leaves us with, I believe, only two options:

1. The solution I tried to implement here - Directly controlling low-level socket options: It is completely portable, it uses standard sockopts, and it is what `actix-web` actually uses under the hood. The only counterargument I can think of is it adds an extra complexity to such a small project.
2. The second option is to not bind to IPv6 by default: Bind only to `0.0.0.0` and the user can always choose to bind to IPv6 by using `-i ::`.
In this way, the default behaviour is reliable across platforms while IPv6 is still a supported feature that the user can use, but the command line argument he should then use depends on his OS and network configuration. A windows user, foe example, shall realize that he needs to bind to both `-i 0.0.0.0 -i ::` in order to support IPv6, and a linux user will quickly realize, with the help of the error messages implemented here, that binding to both addresses is not allowed and binding to `-i ::` is sufficient.

**Comment by aliemjay** [comment]
To implement #422, I pulled `get_if_addrs` crate. It is very simple and cross-platform. The only problem with it is it is no longer maintained!
But I don't see this as a problem because it is only used here for a convenience feature and it is popular enough to trust and has a relatively small code base and it is used here in such a way that any error while getting interface addresses only prints a warning and resumes normally. But I realize that this is controversial.

Other options I have explored are `interfaces`, `nix`, `pnet`, but those are either too heavy or support only unix.
Other less popular options: https://github.com/mensi/netifs-rs , https://github.com/misuzu/rust-interfaces

**Comment by aliemjay** [comment]
> To implement #422, I pulled `get_if_addrs` crate. 
 I tried to use `pnet_datalink`, but unfortunately it turned out it requires libpcap for windows! Reverting this soon...

**Comment by aliemjay** [comment]
To summarize the decision tree regarding the binding behavior issue:

* Bind by default to IPv4 only and let the user deal with the platform-specific behavior of binding to IPv6.
* Or, make consistent cross-platform behavior. In this case the options are:
    * Bind to each interface individually. I mentioned some downsides to this approach previously.
    * Set `ipv6only` sockopt manually (more code and dependencies)
        * Do we set it to `true` or `false` ?

**Comment by aliemjay** [comment]
`README` needs an update!

**Comment by aliemjay** [comment]
@svenstaro I'd appreciate your input here while rebasing this.

File: tests/bind.rs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+#[rstest]
+#[case(&[] as &[&str])]
+#[case(&["-i", "::"])]
+#[case(&["-i", "127.0.0.1"])]
+#[case(&["-i", "0.0.0.0"])]
+#[case(&["-i", "::", "-i", "0.0.0.0"])]
+#[case(&["--random-route"])]
+fn validate_prtinted_urls(tmpdir: TempDir, port: u16, #[case] args: &[&str]) -> Result<(), Error> {
```
Comment (svenstaro) [inline tests/bind.rs:59]: Typo: `validate_printed_urls`

File: tests/bind.rs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+#[case(&["-i", "012.123.234.012"])]
+#[case(&["-i", "::", "-i", "012.123.234.012"])]
+fn bind_fails(tmpdir: TempDir, port: u16, #[case] args: &[&str]) -> Result<(), Error> {
+    Command::cargo_bin("miniserve")?
+        .arg(tmpdir.path())
+        .arg("-p")
+        .arg(port.to_string())
+        .args(args)
+        .stdout(Stdio::null())
+        .stderr(Stdio::null())
```
Comment (svenstaro) [inline tests/bind.rs:23]: Check for the actual error returned here. Not all of it should be a OS error. For instance, it should be possible to check `stderr` for at least `"Failed to bind server"`.

File: src/main.rs [hunk 4]
```diff
@@ -253,6 +230,19 @@ async fn run(miniserve_config: MiniserveConfig) -> Result<(), ContextualError> {
         .map_err(|e| ContextualError::IoError("".to_owned(), e))
 }
 
+/// Allows us to set low-level socket options
```
Comment (svenstaro) [inline src/main.rs:233]: Please add some more comments here with some context as to why we do this, perhaps even link the PR for the discussion in addition as this is a fairly unobvious low-level thing that we have going on here.

**Review by svenstaro — CHANGES_REQUESTED** [review]
A few thoughts in no order:
- Let's order the interface output listing vertically instead of in a long line.
- In your proposed solution 1., do we have to use some OS-specific logic to make the behavior identical between platforms or does it just work?
- I'm generally happy with your solution 1. as I think it's conceptually cleaner with not too much low-level code to maintain in our code base as a result but I'm a little bit worried about compatibility still and we'd definitely have to test it everywhere.
- All in all, I'm pretty happy with the implementation so far and I think we should go ahead with it but it will require some manual testing on the three major OS at least.

**Comment by aliemjay** [comment]
>     * In your proposed solution 1., do we have to use some OS-specific logic to make the behavior identical between platforms or does it just work?

As long as the supported OS's support the IPv6Only option (which all does AFAIK), we don't need to include OS-specific code.

**Comment by svenstaro** [comment]
Please rebase again so that we can see whether these changes build on all target platforms.

**Comment by svenstaro** [comment]
Huh somehow CI doesn't run at all now?

**Comment by svenstaro** [comment]
Ah, probably I need to add back to run it on PR as well.

**Comment by svenstaro** [comment]
I pushed a fix that should do it, please try rebasing again.

**Comment by svenstaro** [comment]
Also, you probably need to update the lock file.

**Comment by svenstaro** [comment]
Looks like it builds fine on all platforms. I'll have access to my Windows machine in a bit and then I should be able to properly test drive this. In the mean time, please also go ahead and test this on as many platforms as you can and see how you like it yourself.

**Comment by svenstaro** [comment]
The tests appear to be timing out with this PR. Can you take a look?

**Comment by svenstaro** [comment]
If you want to rebase again, the changes on master will enable to use easily retrieve all the compiled artifacts from the PR which will allow for much easier cross OS testing. :)

**Comment by svenstaro** [comment]
Evaluating this a bit further, I think we should change the behavior a little bit:
1. Always display to the user what we're binding.
2. Don't advertise the shown IP addresses as being what we're binding but rather provide them as a non-exhaustive list of addresses where miniserve is then available from. Something like
```
Bound to 0.0.0.0, ::1, ::1337
Serving path /tmp
Available at (non-exhaustive list):
- http://10.0.0.1:8080
- http://192.168.1.1:8080
- http://[::1]:8080
- http://127.0.0.1:8080
- http://[::1337]:8080
```
This way, we don't have to lie about users in terms of what we're binding but we can still provide a nice clickable list of output URLs where miniserve is most probably available on. What do you think? If you want to go ahead with this, feel free to change the list format and output to something nicer. This is just what I came up with. :)

**Comment by aliemjay** [comment]
Oh, I've just seen you comment. This is quite reasonable! fixing it now..

File: src/main.rs [hunk 5]
```diff
@@ -240,40 +220,59 @@ async fn run(miniserve_config: MiniserveConfig) -> Result<(), ContextualError> {
...
+
+    println!("Bound to {}", display_sockets.join(", "));
+
+    println!("Serving path {}", Color::Yellow.paint(path_string).bold());
 
     println!(
-        "Serving path {path} at {addresses}",
-        path = Color::Yellow.paint(path_string).bold(),
-        addresses = addresses,
+        "Availabe at (non-exhaustive list):\n\t{}\n",
```
Comment (svenstaro) [inline src/main.rs:247]: I think it'd look better if this were four spaces instead of the 8 spaces tab.

File: src/main.rs [hunk 6]
```diff
@@ -236,10 +239,13 @@ async fn run(miniserve_config: MiniserveConfig) -> Result<(), ContextualError> {
...
+    println!("Bound to {}", display_sockets.join(", "));
+
+    println!("Serving path {}", Color::Yellow.paint(path_string).bold());
+
     println!(
-        "Serving path {path} at:\n\t{addresses}\n",
-        path = Color::Yellow.paint(path_string).bold(),
-        addresses = addresses,
+        "Availabe at (non-exhaustive list):\n    {}\n",
+        display_urls.join("\n    "),
```
Comment (aliemjay) [inline src/main.rs:248]: I wounder if there is a better way to encode spaces?

Comment (svenstaro) [reply src/main.rs:248]: Eh, don't sweat it. We don't need to golf _everything_. :P

**Review by svenstaro — APPROVED** [review]

**Comment by svenstaro** [comment]
Amazing work all in all. It's a fairly hard problem to solve in a cross-platform fashion but I think this is the best possible compromise!

**Pull request merged**